### ISSUE TITLE
Have a repeatable task - never marked as completed

### ISSUE THREAD

**Issue by scdi** [carekit-apple/CareKit#484]
Is is possible to have a repeatable task that is never marked as completed so that a patient may enter some event as often as they experience it in the same day?

**Comment by gavirawson-apple** [comment]
Yep, you can store multiple outcome values on a single outcome for an event. This concept is used by the `OCKButtonLogTaskViewController` if you would like to see a concrete example.

**Comment by scdi** [comment]
Thank you. That is excellent. The task that I have is a survey. When I press the button I get a multistep survey to fill out. Can I do the same thing with OCKButtonLogTaskViewController?

**Comment by majidsaleem105** [comment]
@scdi I had the same condition where patients can submit a few surveys multiple time so I resolved it this way:

I created a custom surveys class like:

`class SurveyViewController: OCKInstructionsTaskViewController, ORKTaskViewControllerDelegate {

func taskViewController(_ taskViewController: ORKTaskViewController, didFinishWith reason: ORKTaskViewControllerFinishReason, error: Error?) {
        
        guard reason == .completed else {
            taskView.completionButton.isSelected = ( resubmittedSurveys.contains( (surveyTask?.title)! ) ? true : false)
            return
        }
}
}`

Where `resubmittedSurveys` is an array of survey names that can be submitted multiple times.

To create surveys you can see the last reply from @erik-apple on https://github.com/carekit-apple/CareKit/issues/401#event-3343058734

**Comment by scdi** [comment]
Thank you so so much. I will give that a try.

**Comment by majidsaleem105** [comment]
@scdi 

I forgot to add the following code in my last reply. In the updateView function you can see whether the survey can be submitted multiple times are not so based on your logic you can add the following line to make the Completion button active:
`view.completionButton.isUserInteractionEnabled = ( resubmittedSurveys.contains( surveyTitle! ) ? true : false)`

class SurveyViewSynchronizer: OCKInstructionsTaskViewSynchronizer {
    
    // Customize the initial state of the view
    override func makeView() -> OCKInstructionsTaskView {
        let instructionsView = super.makeView()
        instructionsView.completionButton.label.text = "Start Survey"
        return instructionsView
    }
    
    // Customize how the view updates
    override func updateView(_ view: OCKInstructionsTaskView,
                             context: OCKSynchronizationContext<OCKTaskEvents?>) {
        super.updateView(view, context: context)
        
        // Check if an answer exists or not and set the detail label accordingly
        if (context.viewModel?.firstEvent?.outcome?.values.first) != nil {
            view.headerView.detailLabel.text = "" //context.viewModel?.firstEvent?.task.title //"Pain Rating: \(answer)"
            
            var surveyTitle = (context.viewModel?.firstEvent?.task.title!)
            surveyTitle = surveyTitle!.replacingOccurrences(of: "(New) ", with: "")

            view.completionButton.isUserInteractionEnabled = ( resubmittedSurveys.contains( surveyTitle! ) ? true : false)
        } else {

            view.headerView.detailLabel.text = "" //"Rate your pain on a scale of 1 to 10"
        }
    }
}

**Comment by scdi** [comment]
Yeah, thank you. I thought I was missing something. I will try that. Thanks so much.

**Comment by erik-apple** [comment]
@scdi Were you able to get this one figured out?

**Comment by scdi** [comment]
I tried it immediately but there was a problem that I can't remember and thought I would get back to it to troubleshoot. I will try again.

**Comment by scdi** [comment]
@erik-apple, thank you for circling back. I am attaching a file of the implementation that I have.  That could work. I have a few hick-ups. 
1. I wanted to have double values on the graph but I was only been able to get integers.
2. When pressing the "Completed" button returning to SCD Daily Check-in to allow another entry, the graph data is cleared.
3. When 0/10 is entered it is as if there were no entries on that day, I wish there was a way to indicate that the value entered was zero. 
I think that items 3 might be for the ResearchKit team.

Altogether, I am very happy with we are able to do so far.
This CK/RK is a very elegant and practical set of frameworks.


![Screen Recording 2020-08-14 at 3 47 36 PM](https://user-images.githubusercontent.com/13545391/90293418-01606300-de52-11ea-9c83-eddf567b4ceb.gif)

**Comment by erik-apple** [comment]
> 1. I wanted to have double values on the graph but I was only been able to get integers.
> 2. When pressing the "Completed" button returning to SCD Daily Check-in to allow another entry, the graph data is cleared.
> 3. When 0/10 is entered it is as if there were no entries on that day, I wish there was a way to indicate that the value entered was zero.

Let's tackle those one at a time. Getting double values is pretty straight forward. I presume when you convert the value collected by ResearchKit into an `OCKOutcomeValue`, you are storing it as a double? Begin by confirming that please.

Then we'll need to setup a custom event aggregator to pass into the chart. Have you do that already? It'll look something like this...

```swift
let myAggregator = OCKEventAggregator.custom({ oneDayOfEvents -> Double in 
    var total = 0.0
    for event in oneDayOfEvents {
        if let value = event.outcome?.values?.first?.doubleValue {
            total += value
        }
    }
    return total
})
```

Then you pass the aggregator into the `OCKDataSeriesConfiguration` that you display in your chart.

We can address the other two questions once we get past this first one.

**Comment by scdi** [comment]
Thank you so  much for looking into this. The problem is 
event.outcome?.values is always returning an array of integer [2, 5, etc] values rather than doubles [2.3, 5.1, etc.] in

let myAggregator = OCKEventAggregator.custom({ oneDayOfEvents -> Double in 
    var total = 0.0
    for event in oneDayOfEvents {
        if let value = **event.outcome?.values**?.first?.doubleValue {
            total += value
        }
    }
    return total
})

I have the answer format for the step as: 

let answerFormat = ORKAnswerFormat.continuousScale(withMaximumValue: 10.0, minimumValue: 0.0, defaultValue: 0.0, maximumFractionDigits: 1, vertical: false, maximumValueDescription: "worst pain", minimumValueDescription: "no pain")

We still get back Integers

**Comment by scdi** [comment]
For clarification, the arrays returned contain only one value, e.g., [2].

**Comment by scdi** [comment]
I am sure if this will help but if we try to get a Double value in **event.outcome?.values** we get this error:
**"Initializer 'init(_:)' requires that 'OCKOutcomeValue' conform to 'BinaryInteger'"**

**Comment by erik-apple** [comment]
Can you post a snippet of the code you are using to save the outcome to CareKit after completing the ResearchKit survey? I suspect that there is a type conversion mismatch somewhere in there!

**Comment by scdi** [comment]
Hi, I was saving the value as an Int. I changed it to 
Retrieve the result from the ResearchKit survey
        let survey = taskViewController.result.results!.first(where: { $0.identifier == "_scd_pain" }) as! ORKStepResult
        let painResult = survey.results!.first as! ORKScaleQuestionResult
        let answer =  painResult.scaleAnswer!

Now I am able to data as Double.

**Comment by erik-apple** [comment]
Great! I'm glad you got it working!

Your second question was:
> 2. When pressing the "Completed" button returning to SCD Daily Check-in to allow another entry, the graph data is cleared.

Does it clear the data for just the currently selected day, or all days?

**Comment by scdi** [comment]
Thanks for your help. In item 2. the data  cleared only for that day.

**Comment by erik-apple** [comment]
Can you show a snippet of the code where you handle dismissal of the ResearchKit view controller? I suspect you're accidentally overwriting the existing value with an empty one.

**Comment by scdi** [comment]
func taskViewController(_ taskViewController: ORKTaskViewController, didFinishWith reason: ORKTaskViewControllerFinishReason, error: Error?) {
        **taskViewController.dismiss(animated: true, completion: nil)**
 
        guard reason == .completed else {
            taskView.completionButton.isSelected = false
            return
        }
        
        // Retrieve the result from the ResearchKit survey
        let survey = taskViewController.result.results!.first(where: { $0.identifier == "_scd_pain" }) as! ORKStepResult
        let painResult = survey.results!.first as! ORKScaleQuestionResult
        let answer = painResult.scaleAnswer!
        print("all pain results - \(String(describing: survey.results))")
       
 // 4b. Save the result into CareKit's store
        controller.appendOutcomeValue(value: answer as! Double, at: IndexPath(item: 0, section: 0), completion: nil)

       ......


![RPReplay_Final1597778187](https://user-images.githubusercontent.com/13545391/90556300-adfa5780-e166-11ea-8b7f-5f6b09d4d0fa.gif)

**Comment by erik-apple** [comment]
Oh, I think I understand now. You would like it not to delete the answer when you tap the completed button!
That happens in your subclass. Try commenting out step 2a.

```swift
class SurveyViewController: OCKInstructionsTaskViewController, ORKTaskViewControllerDelegate {

    // 2. This method is called when the use taps the button!
    override func taskView(_ taskView: UIView & OCKTaskDisplayable, didCompleteEvent isComplete: Bool, at indexPath: IndexPath, sender: Any?) {

        // Try commenting this code out! This is where the existing outcome gets deleted.
        //
        // 2a. If the task was marked incomplete, fall back on the super class's default behavior (deleting the outcome).
        // if !isComplete {
        //    super.taskView(taskView, didCompleteEvent: isComplete, at: indexPath, sender: sender)
        //    return
        // }

        // 2b. If the user attempted to mark the task complete, display a ResearchKit survey.
        let answerFormat = ORKAnswerFormat.scale(withMaximumValue: 10, minimumValue: 1, defaultValue: 5, step: 1, vertical: false,
                                                 maximumValueDescription: "Very painful", minimumValueDescription: "No pain")
        let painStep = ORKQuestionStep(identifier: "pain", title: "Pain Survey", question: "Rate your pain", answer: answerFormat)
        let surveyTask = ORKOrderedTask(identifier: "survey", steps: [painStep])
        let surveyViewController = ORKTaskViewController(task: surveyTask, taskRun: nil)
        surveyViewController.delegate = self

        // 3a. Present the survey to the user
        present(surveyViewController, animated: true, completion: nil)
    }
}
```

**Comment by scdi** [comment]
Wow. CareKit is amazing and the team rocks.

**Comment by erik-apple** [comment]
We do our best!

> 3. When 0/10 is entered it is as if there were no entries on that day, I wish there was a way to indicate that the value entered was zero.

Can you elaborate on this a bit more? Do you want to add an indication inside of the ResearchKit view controller, or on CareKit's chart?

**Comment by scdi** [comment]
When I look at the graph for a day when the value zero is entered and a day when no value are entered, the y-axis does not capture a difference. It would be nice if when a value is missing, the day of the week is colored orange or blue or have an x on the x-axis to show that data entry was missed on that day. The completion circle help somewhat but it would nice to have a give the feedback that pain was zero vs no pain value was entered on that day.
Please compare Monday and Wednesday on the graph. That is what one would see if on Wednesday there is no value entered.
![Screenshot 2020-08-18 at 6 42 55 PM](https://user-images.githubusercontent.com/13545391/90573042-6a621680-e183-11ea-8be7-fffa797e0e15.jpeg)

**Comment by erik-apple** [comment]
I see what you're getting at.

Perhaps you could do something like showing 0.1 for 0 to give a small visual indication that an answer was given, but still make it clear that the answer was 0.

```swift
let myAggregator = OCKEventAggregator.custom({ oneDayOfEvents -> Double in 
    
    // Add up all the answers that were given
    let answers = oneDayOfEvents.compactMap { $0.outcome?.values?.first?.doubleValue }
    var total = answers.reduce(0, +)
    
    // If answers were given, but they were all 0
    if !answers.isEmpty && total == 0 {
        total = 0.1
    }

    return total
})
```

Alternatively, you could try out line or scatter plots. Those chart styles are both provide more visual distinction between "no answer" and "0".

**Comment by marco-theraforge** [comment]
To give a more immediate and intuitive visual guidance, would it be possible, say, to grey out the corresponding day's label and/or to add an "N/A" visual indication in the graph for the days when no value is entered?

**Comment by marco-theraforge** [comment]
A visual indication like in this graph:

![image](https://user-images.githubusercontent.com/52140220/90575586-04de4c00-e1bc-11ea-9545-aad6499d6261.jpeg)
![image](https://user-images.githubusercontent.com/52140220/90575589-0576e280-e1bc-11ea-9eb8-9d4f914ad9d3.png)

**Comment by erik-apple** [comment]
There isn't a way to do that today. It would require PR'ing a new feature into CareKit.

**Comment by scdi** [comment]
I use 0.2 instead of 0.1 and I like the fact that it can be seen. The result is a workable option. Thank you for looking into this. Could we gray out just the letter representing a day without data entry? Maybe that could be done. 
![IMG_0008](https://user-images.githubusercontent.com/13545391/90586057-47942a00-e1a4-11ea-90f3-952f109f2e7a.jpeg)
